### ISSUE TITLE
Single selection and ActiveSelection don't react when object becomes not selectable

### ISSUE THREAD

**Issue by kirill-konshin** [fabricjs/fabric.js#7608]
## Version
4.5.0

## Test Case

http://jsfiddle.net/dfuse/cyneb8rg/

## Steps to reproduce

First scenario, the simple one:

1. User selects single / multiple objects
2. Some code (click on `Unselect` button) sets `selected:false` (and `stroke:red` for visual clue)

https://user-images.githubusercontent.com/1530410/149210708-1132a608-252c-474a-a619-d2eff2a20b72.mp4

Second, a bit more complicated scenario:

1. User selects single / multiple objects
2. User starts to move objects around
3. Some code (`setTimeout` for sake of demo) sets `selected:false` (and `stroke:red` for visual clue)

## Expected Behavior

Fabric should track `selectable` property and if it becomes `false` remove object from selection (scenario 1 and 2) and abort transformation (skew, scale, move, rotate) when single or multiple objects are transforming (scenario 2).

## Actual Behavior

Selection is unchanged (scenario 1), transformation continues (scenario 2), object becomes unselectable only after selection is discarded (click anywhere on canvas for example).

Note that rect becomes red but nothing happens.

## Proposed solution

I went ahead and created a smarter deselector `makeUnselectableSmart`: 

```js
const oldObjects = canvas.getActiveObjects();
const objects = oldObjects.filter(o => o !== rect);

if (objects.length === oldObjects.length) return;

canvas.discardActiveObject().requestRenderAll();
  
if (!objects.length) return;

const selection = new fabric.ActiveSelection(objects, { canvas });

canvas.setActiveObject(selection).requestRenderAll();
```

I suppose this kind of boilerplate code can become part of lib... or a default behavior.

@asturur @ShaMan123 pls take a look.

**Comment by ShaMan123** [comment]
I gave this a quick read.
So I may have misunderstood.
fabric doesn't work with setters (as you saw #7596 it may will)
So of course nothing happens once you change the `selected` property.
Props not really change state.
You need to call so method.
I agree it is not desirable, old fashioned.
Think of PRing.
And take note that ActiveSelection will probably be rewritten too (it is a subclass of group)

**Comment by kirill-konshin** [comment]
Yep, I'm using `set` method.

The problem with my solution is if I drag single object, timeout kicks in, even with `makeUnselectableSmart` (code above) the object is not unselected: `canvas.discardActiveObject()` has no effect when single object.

https://user-images.githubusercontent.com/1530410/149360575-e8bf5f85-1e4b-4e14-a78a-3ccd7e91f65b.mp4

As you see when I drag the object it becomes red (at this moment `rect.set({selectable: false})` and `makeUnselectableSmart` are executed, but I still can drag it around.

**Comment by ShaMan123** [comment]
I wonder if the solution lies in the event handler.
A simple check there should do the job.
Look here:
https://github.com/fabricjs/fabric.js/blob/9d8fce4a568efc1dafabaceb69017aba30777453/src/mixins/canvas_events.mixin.js#L833
```diff
- else {
+ else if(this._currentTransform.target.selectable) {
```

**Comment by kirill-konshin** [comment]
@ShaMan123 can you produce a build so I can verify?

**Comment by ShaMan123** [comment]
I suggest you fork, create a branch and commit the change.
Then use the cdn you were using pointing to the branch if you need it in fiddle.

**Comment by asturur** [comment]
I gave a look to the current fiddles but i couldn't understand which is the final use case of the issue.
I understand you want something that will auto deselect stuff based on something.
My gut answer is no, we are missing something.

If setting `selectable = false` needs to deselect something, since the user has no way to set selectable property with mouse or keyboard, it means your code is setting it.
If your code does it, it can do something else to actually deselect the object.

Can you explain a bit more your usecase for this feature, like which is the application feature you are trying to build.

**Comment by kirill-konshin** [comment]
@asturur like I said earlier, the object may become non-interactable during interaction (due to race conditions and other causes), so we need to abort transformation immediately and deselect an object.

My fiddle http://jsfiddle.net/dfuse/cyneb8rg/ allows to abort transformation when there are multiple objects selected, but when it's only one object, it can't be aborted, or I don't know how to abort it.

**Comment by ShaMan123** [comment]
> ```diff
> else if(this._currentTransform.target.selectable) {
> ```

Seems to work
http://jsfiddle.net/symvad17/4/

I think that changing the selectable prop is confusing because it is already selected so I would expect that not make a differnce but instead some other flag like `lockMovementX`.
In general I think what's really needed is a mechanism on top of events that allows the dev to do exactly this by returning a boolean value.

**Comment by ShaMan123** [comment]
I see the other action handler do check the `lock` flags and abort accordingly.
So just a simple `lockMovement` check is needed

**Comment by ShaMan123** [comment]
This can be closed in my opinion

**Comment by kirill-konshin** [comment]
> This can be closed in my opinion

Is it fixed? :) 

I mean that line is committed and released?

> lockMovementX lockMovementY do exactly what you want

Also probably `lockRotation`, `lockScalingX`, `lockScalingY`, `lockSkewingX` and `lockSkewingY`. And also `selectable` and `evented`. All set to `false`.

**Comment by kirill-konshin** [comment]
> Seems to work without the need to workaround
> http://jsfiddle.net/symvad17/8/

Not quite, selection still has to be dropped manually, e.g. object stops moving, but remains selected

**Comment by ShaMan123** [comment]
Yes, you can/should call `discardActiveObjects`
